core/device.cpp

```diff
@@ -197,3 +199,102 @@ extern "C" void free_device_table(struct device_table *devices)
 {
 	delete devices;
 }
+
+// managing fingerprint data
+bool fingerprint_record::operator<(const fingerprint_record &a) const
+{
+	if (model == a.model)
+		return serial < a.serial;
+	return model < a.model;
+}
+
+// annoyingly, the Cressi Edy doesn't support a serial number (it's always 0), but still uses fingerprints
+// so we can't bail on the serial number being 0
+extern "C" unsigned int get_fingerprint_data(const struct fingerprint_table *table, uint32_t model, uint32_t serial, unsigned char **fp_out)
+{
+	if (model == 0 || fp_out == nullptr)
+		return 0;
+	struct fingerprint_record fpr = { model, serial };
+	auto it = std::lower_bound(table->fingerprints.begin(), table->fingerprints.end(), fpr);
+	if (it != table->fingerprints.end()) {
```

Comment:
std::lower_bound finds the first element that is greater or equal. Therefore, you probably have to check for equality as well?

Reply:
Yes - that's a clear bug in the code.
Teaches me not to use these fancy functions that I don't really understand, anyway.